wgpu-core/src/binding_model.rs

```diff
@@ -446,25 +446,64 @@ pub type BindGroupLayouts<A> = crate::storage::Storage<BindGroupLayout<A>, BindG
 ///  - produced bind groups
 ///  - produced pipeline layouts
 ///  - pipelines with implicit layouts
-#[derive(Debug)]
 pub struct BindGroupLayout<A: hal::Api> {
-    pub(crate) raw: A::BindGroupLayout,
     pub(crate) device_id: Stored<DeviceId>,
     pub(crate) multi_ref_count: MultiRefCount,
-    pub(crate) entries: BindEntryMap,
     // When a layout created and there already exists a compatible layout the new layout
     // keeps a reference to the older compatible one. In some places we substitute the
     // bind group layout id with its compatible sibling.
     // Since this substitution can come at a cost, it is skipped when wgpu-core generates
     // its own resource IDs.
-    pub(crate) compatible_layout: Option<Valid<BindGroupLayoutId>>,
+    pub(crate) inner: BglOrDuplicate<A>,
+}
+
+pub(crate) enum BglOrDuplicate<A: hal::Api> {
+    Inner(BindGroupLayoutInner<A>),
+    Duplicate(Valid<BindGroupLayoutId>),
+}
```

Comment:
nitpick: I think this naming might be clearer:

```suggestion
pub(crate) enum BindGroupLayoutDedupedInner<A: hal::Api> {
    Original(BindGroupLayoutInner<A>),
    Duplicate(Valid<BindGroupLayoutId>),
}
```

…with according renames for methods, of course.

Reply:
I prefer the current name. It's sandwitched between `BindGroupLayout` and `BindGroupLayoutInner` so in every context where it is used there is zero ambiguity about what "Bgl" means. The enum represents either a real BGL or a duplicate indirection, so the "A or B" formulation seems pretty clear to me (I'm not sure how to interpret `DedupedInner` in your proposal).

Reply:
🤷🏻‍♂️Not concerned either way. Onward!